test/auth_aws/test_auth_oidc.py

```diff
@@ -0,0 +1,785 @@
+        refresh_cb = self.create_refresh_cb()
+
+        # Create a client with the callbacks.
+        props: Dict = dict(request_token_callback=request_cb, refresh_token_callback=refresh_cb)
+        client = MongoClient(self.uri_single, authmechanismproperties=props)
+
+        # Perform a find operation.
+        client.test.test.find_one()
+
+        # Perform a bulk write operation.
+        client.test.test.bulk_write([InsertOne({})])
+
+        # Assert that the refresh callback has not been called.
+        self.assertEqual(self.refresh_called, 0)
+
+        with self.fail_point(
+            {
+                "mode": {"times": 1},
+                "data": {"failCommands": ["find"], "errorCode": 391},
+            }
+        ):
+            # Perform a bulk read operation.
+            cursor = client.test.test.find_raw_batches({})
+            list(cursor)
+
+        # Assert that the refresh callback has been called.
+        self.assertEqual(self.refresh_called, 1)
+        client.close()
+
+    def test_reauthenticate_succeeds_cursor(self):
+        request_cb = self.create_request_cb()
+        refresh_cb = self.create_refresh_cb()
+
+        # Create a client with the callbacks.
+        props: Dict = dict(request_token_callback=request_cb, refresh_token_callback=refresh_cb)
+        client = MongoClient(self.uri_single, authmechanismproperties=props)
+
+        # Perform an insert operation.
+        client.test.test.insert_one({"a": 1})
+
+        # Assert that the refresh callback has not been called.
+        self.assertEqual(self.refresh_called, 0)
+
+        with self.fail_point(
+            {
+                "mode": {"times": 1},
+                "data": {"failCommands": ["find"], "errorCode": 391},
+            }
+        ):
+            # Perform a find operation.
+            cursor = client.test.test.find({"a": 1})
+            self.assertGreaterEqual(len(list(cursor)), 1)
+
+        # Assert that the refresh callback has been called.
+        self.assertEqual(self.refresh_called, 1)
+        client.close()
+
+    def test_reauthenticate_succeeds_get_more(self):
+        request_cb = self.create_request_cb()
+        refresh_cb = self.create_refresh_cb()
+
+        # Create a client with the callbacks.
+        props: Dict = dict(request_token_callback=request_cb, refresh_token_callback=refresh_cb)
+        client = MongoClient(self.uri_single, authmechanismproperties=props)
+
+        # Perform an insert operation.
+        client.test.test.insert_many([{"a": 1}, {"a": 1}])
+
+        # Assert that the refresh callback has not been called.
+        self.assertEqual(self.refresh_called, 0)
+
+        with self.fail_point(
+            {
+                "mode": {"times": 1},
+                "data": {"failCommands": ["find"], "errorCode": 391},
+            }
+        ):
+            # Perform a find operation.
+            cursor = client.test.test.find({"a": 1}, batch_size=1, cursor_type=CursorType.EXHAUST)
```

Comment:
getMore should be tested with both regular cursors and exhaust cursors. Also exhaust needs @client_context.require_no_mongos.

Reply:
Done